### PR TITLE
Generic gen valid

### PR DESCRIPTION
**Pull request by NorfairKing**
@CSVdB This is an attempt at implementing #39, which was your idea. 
I'd like to request your review.

### CODE REVIEW

File: genvalidity/src/Data/GenValidity.hs [hunk 1]
```diff
@@ -115,13 +115,13 @@ class GenUnchecked a where
     genUnchecked :: Gen a
     default genUnchecked :: (Generic a, GGenUnchecked (Rep a)) =>
         Gen a
-    genUnchecked = to <$> gGenUnchecked
+    genUnchecked = genericGenUnchecked
```
Comment (CSVdB) [inline genvalidity/src/Data/GenValidity.hs]: Renames shouldn't be in the same PR as an added feature.

File: genvalidity/src/Data/GenValidity.hs [hunk 2]
```diff
@@ -115,13 +115,13 @@ class GenUnchecked a where
...
 
     shrinkUnchecked :: a -> [a]
     default shrinkUnchecked ::
         (Generic a, GUncheckedRecursivelyShrink (Rep a), GUncheckedSubterms (Rep a) a) =>
         a -> [a]
-    shrinkUnchecked = gShrinkUnchecked
+    shrinkUnchecked = genericShrinkUnchecked
```
Comment (CSVdB) [inline genvalidity/src/Data/GenValidity.hs]: Idem

File: genvalidity/src/Data/GenValidity.hs [hunk 3]
```diff
@@ -625,6 +625,9 @@ genListOf func =
         pars <- arbPartition size
         forM pars $ \i -> resize i func
 
+genericGenUnchecked :: (Generic a, GGenUnchecked (Rep a)) => Gen a
+genericGenUnchecked = to <$> gGenUnchecked
```
Comment (CSVdB) [inline genvalidity/src/Data/GenValidity.hs]: Idem

File: genvalidity/src/Data/GenValidity.hs [hunk 4]
```diff
@@ -735,3 +738,122 @@ instance OVERLAPPING_ GUncheckedSubtermsIncl (K1 i a) a where
 instance OVERLAPPING_ GUncheckedSubtermsIncl (K1 i a) b where
   gUncheckedSubtermsIncl (K1 _) = []
 
+
+genValidStructurally :: (Validity a, Generic a, GGenValid (Rep a)) => Gen a
+genValidStructurally = genValidStructurallyWithoutExtraChecking `suchThat` isValid
+
+genValidStructurallyWithoutExtraChecking :: (Generic a, GGenValid (Rep a)) => Gen a
```
Comment (CSVdB) [inline genvalidity/src/Data/GenValidity.hs]: Good idea, since checking can be slow!

File: genvalidity/test/Data/GenValidity/GenericSpec.hs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+import Data.Proxy
+import Data.Typeable
+
+import Data.GenValidity
+
+spec :: Spec
+spec = do
+    describe "genValidStructurally" $ do
+        genValidstructurallySpec (Proxy :: Proxy Bool)
```
Comment (CSVdB) [inline genvalidity/test/Data/GenValidity/GenericSpec.hs]: Why Proxy instead of TypeApplications?

Comment (NorfairKing) [reply genvalidity/test/Data/GenValidity/GenericSpec.hs]: because it needs to work with ghc-7.8

File: genvalidity/src/Data/GenValidity.hs (hunk 1 above)
Comment (NorfairKing) [reply genvalidity/src/Data/GenValidity.hs]: Good point.

**Comment by CSVdB** [comment]
Looks great, good work!
Will be merged once travis passes.

**Pull request merged**